sdk/python/kfp/v2/dsl/container_op.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""IR-based ContainerOp."""
+
+from typing import Callable, Text
+
+from kfp import dsl
+from kfp.v2.proto import pipeline_spec_pb2
+
+# Unit constants for k8s size string.
+_E = 10 ** 18  # Exa
+_EI = 1 << 60  # Exa: power-of-two approximate
+_P = 10 ** 15  # Peta
+_PI = 1 << 50  # Peta: power-of-two approximate
+# noinspection PyShadowingBuiltins
+_T = 10 ** 12  # Tera
+_TI = 1 << 40  # Tera: power-of-two approximate
+_G = 10 ** 9  # Giga
+_GI = 1 << 30  # Giga: power-of-two approximate
+_M = 10 ** 6  # Mega
+_MI = 1 << 20  # Mega: power-of-two approximate
+_K = 10 ** 3  # Kilo
+_KI = 1 << 10  # Kilo: power-of-two approximate
+
+_GKE_ACCELERATOR_LABEL = 'cloud.google.com/gke-accelerator'
+
+# Shorthand for PipelineContainerSpec
+_PipelineContainerSpec = pipeline_spec_pb2.PipelineDeploymentConfig.PipelineContainerSpec
+
+
+def resource_setter(func: Callable):
+  """Function decorator for common validation before setting resource spec."""
+
+  def resource_setter_wrapper(container_op: 'ContainerOp', *args,
+      **kwargs) -> 'ContainerOp':
+    # Validate the container_op has right format of container_spec set.
+    if not hasattr(container_op, 'container_spec'):
+      raise ValueError('Expecting container_spec attribute of the container_op:'
+                       ' {}'.format(container_op))
+    if not isinstance(
+        container_op.container_spec, _PipelineContainerSpec):
+      raise TypeError('ContainerOp.container_spec is expected to be a '
+                      'PipelineContainerSpec proto. Got: {} for {}'.format(
+          type(container_op.container_spec), container_op.container_spec))
+    # Run the resource setter function
+    return func(container_op, *args, **kwargs)
+
+  return resource_setter_wrapper
+
+
+def _get_cpu_number(cpu_string: Text) -> float:
+  """Converts the cpu string to number of vCPU core."""
+  # dsl.ContainerOp._validate_cpu_string guaranteed that cpu_string is either
+  # 1) a string can be converted to a float; or
+  # 2) a string followed by 'm', and it can be converted to a float.
+  if cpu_string.endswith('m'):
+    return float(cpu_string[:-1]) / 1000
+  else:
+    return float(cpu_string)
+
+
+def _get_memory_number(memory_string: Text) -> float:
```

Comment:
This might be used for resources other than memory (e.g. disks).

Reply:
Renamed.